src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -2026,13 +2244,132 @@ class ArnoldLight : public ArnoldObject
 		NodeDeleter m_nodeDeleter;
 		const AtNode *m_parentNode;
 		ArnoldShaderPtr m_lightShader;
+		BlockerConnections *m_connections;
+		std::vector<ArnoldShaderPtr> m_lightFilterShaders;
 
 };
 
 IE_CORE_DECLAREPTR( ArnoldLight )
 
 } // namespace
 
+
+namespace
+{
+
+void BlockerConnections::registerLight( const std::string &lightName, ArnoldLight *light )
+{
+	ConnectionsMap::accessor a;
+	m_connections.insert( a, lightName );
+
+	a->second.light = light;
+	a->second.dirty = true;
+
+	if( m_ownConnections )
+	{
+		m_arnoldObjects.emplace_back( light );
+	}
+}
+
+void BlockerConnections::deregisterLight( const std::string &lightName )
+{
+
+	ConnectionsMap::accessor a;
+	if( m_connections.find( a, lightName ) )
+	{
+		a->second.light = nullptr;
+		a->second.dirty = true;
+	}
+	else
+	{
+		throw IECore::Exception( boost::str( boost::format( "Can not deregister non-existing light \"%s\"" ) % lightName.c_str() ) );
+	}
+}
+
+void BlockerConnections::registerBlocker( const std::vector<std::string> &lightNames, ArnoldBlocker *blocker )
+{
+	for( const std::string &lightName : lightNames )
+	{
+		ConnectionsMap::accessor a;
+		if( !m_connections.find( a, lightName ) )
+		{
+			throw IECore::Exception( boost::str( boost::format( "Unable to register blocker for non-existing light \"%s\"" ) % lightName.c_str() ) );
+		}
+
+		a->second.blockers.insert( blocker );
+
+		// Even if we knew about the blocker already we need to dirty the light
+		// as the blocker itself might have been updated.
+		a->second.dirty = true;
+	}
+
+	if( m_ownConnections )
+	{
+		m_arnoldObjects.emplace_back( blocker );
+	}
+}
+
+void BlockerConnections::deregisterBlocker( ArnoldBlocker *blocker )
+{
+	for( auto &entry : m_connections )
```

Comment:
`concurrent_hash_map` doesn't allow concurrent iteration, and iterating over _all_ lights seems like more work than we want to do anyway. Can we instead just pass in the lights that we want to deregister ourselves from?